File: sdk/android/src/main/kotlin/io/quicp/QuicpEngine.kt
package io.quicp

import java.net.Inet6Address
import java.net.InetAddress
import java.nio.ByteBuffer

data class QuicpSocketAddress(val address: String, val port: Int)
data class QuicpPath(val local: QuicpSocketAddress, val peer: QuicpSocketAddress)
enum class QuicpStatus(val wire: Int) {
    OK(0), WOULD_BLOCK(1), BUFFER_TOO_SMALL(2), INVALID_ARGUMENT(3),
    NOT_READY(4), CLOSED(5), PANIC(6), FAILED(7);

    companion object {
        fun decode(wire: Int): QuicpStatus =
            entries.firstOrNull { it.wire == wire }
                ?: throw IllegalStateException("unknown QUICP status $wire")
    }
}
data class QuicpIoResult(val status: QuicpStatus, val bytes: Int)
data class QuicpDriveResult(val status: QuicpStatus, val tasks: Int)
data class QuicpRecoverySnapshot(
    val sourceSent: Long,
    val sourceReceived: Long,
    val repairSent: Long,
    val recovered: Long,
    val replayed: Long,
    val fallback: Long,
    val dropped: Long,
    val earlyAccepted: Long,
    val earlyRejected: Long,
    val pathLostPackets: Long,
    val maxPathRttMicros: Long,
    val queuedDatagrams: Long,
    val retainedSourceBytes: Long,
)

/** Single-owner synchronous engine. The owning event loop serializes every call. */
class QuicpEngine private constructor(private var handle: Long) : AutoCloseable {
    enum class Role(val wire: Int) { CLIENT(1), SERVER(2) }
    enum class RecoveryMode(val wire: Int) { ADAPTIVE(1), RELIABLE_ONLY(2) }
    sealed interface Security {
        data object Insecure : Security
        data class MutualTls(
            val serverName: String,
            val caCertificate: String,
            val certificate: String,
            val privateKey: String,
        ) : Security
    }

    companion object {
        init { System.loadLibrary("quicp_jni") }

        @JvmStatic
        fun create(
            role: Role,
            paths: List<QuicpPath>,
            packetCapacity: Int = 256,
            mtu: Int = 1500,
            recoveryMode: RecoveryMode = RecoveryMode.ADAPTIVE,
            security: Security = Security.Insecure,
        ): QuicpEngine {
            require(paths.size in 1..2)
            require(packetCapacity > 0 && mtu > 0)
            val config = ByteBuffer.allocateDirect(nativeConfigSize())
            check(
                status(
                    nativeInitConfig(
                        config, role.wire, paths.size, packetCapacity, mtu, recoveryMode.wire
                    )
                ) == QuicpStatus.OK
            )
            paths.forEachIndexed { index, path ->
                check(
                    status(
                        nativeSetPath(
                            config, index,
                            address(path.local), path.local.port,
                            address(path.peer), path.peer.port,
                        )
                    ) == QuicpStatus.OK
                )
            }
            val status = IntArray(1)
            val handle = when (security) {
                Security.Insecure -> nativeCreate(config, status)
                is Security.MutualTls -> nativeCreateTls(
                    config,
                    status,
                    security.serverName.encodeToByteArray(),
                    security.caCertificate.encodeToByteArray(),
                    security.certificate.encodeToByteArray(),
                    security.privateKey.encodeToByteArray(),
                )
            }
            if (handle == 0L) throw QuicpException(QuicpStatus.decode(status[0]))
            return QuicpEngine(handle)
        }

        private fun address(value: QuicpSocketAddress): ByteArray {
            require(value.port in 1..65535)
            return if (value.address.contains(':')) {
                require(!value.address.contains('%'))
                val parsed = InetAddress.getByName(value.address)
                require(parsed is Inet6Address)
                parsed.address
            } else {
                val octets = value.address.split('.')
                require(octets.size == 4)
                octets.map { octet ->
                    require(octet.isNotEmpty() && (octet == "0" || !octet.startsWith('0')))
                    require(octet.all(Char::isDigit))
                    val number = octet.toInt()
                    require(number in 0..255)
                    number.toByte()
                }.toByteArray()
            }
        }

        @JvmStatic private external fun nativeConfigSize(): Int
        @JvmStatic private external fun nativeInitConfig(
            config: ByteBuffer, role: Int, pathCount: Int, packetCapacity: Int,
            mtu: Int, recoveryMode: Int,
        ): Int
        @JvmStatic private external fun nativeSetPath(
            config: ByteBuffer, path: Int,
            localAddress: ByteArray, localPort: Int,
            peerAddress: ByteArray, peerPort: Int,
        ): Int
        @JvmStatic private external fun nativeCreate(config: ByteBuffer, status: IntArray): Long
        @JvmStatic private external fun nativeCreateTls(
            config: ByteBuffer,
            status: IntArray,
            serverName: ByteArray,
            caCertificate: ByteArray,
            certificate: ByteArray,
            privateKey: ByteArray,
        ): Long
        @JvmStatic private external fun nativeDrive(handle: Long, elapsedNanos: Long, maxTasks: Int): Long
        @JvmStatic private external fun nativeNextTimer(handle: Long): Long
        @JvmStatic private external fun nativeConnectionState(handle: Long): Int
        @JvmStatic private external fun nativeRecoverySnapshot(handle: Long, output: LongArray): Int
        @JvmStatic private external fun nativeIngress(handle: Long, path: Int, data: ByteBuffer, length: Int): Int
        @JvmStatic private external fun nativeEgress(handle: Long, path: Int, output: ByteBuffer, limit: Int): Long
        @JvmStatic private external fun nativePathUnavailable(handle: Long, path: Int): Int
        @JvmStatic private external fun nativeOpenFlow(
            handle: Long, host: ByteArray, port: Int, flow: LongArray,
        ): Int
        @JvmStatic private external fun nativeOpenReplaySafeFlow(
            handle: Long,
            token: ByteBuffer,
            tokenLength: Int,
            nonce: Long,
            host: ByteArray,
            port: Int,
            initial: ByteBuffer,
            initialLength: Int,
            flow: LongArray,
        ): Int
        @JvmStatic private external fun nativePollFlowRequest(
            handle: Long,
            replay: Boolean,
            host: ByteBuffer,
            initial: ByteBuffer,
            metadata: LongArray,
        ): Int
        @JvmStatic private external fun nativeAcceptPendingFlow(
            handle: Long, request: Long, flow: LongArray,
        ): Int
        @JvmStatic private external fun nativeRejectPendingFlow(handle: Long, request: Long): Int
        @JvmStatic private external fun nativeConfigureReplayAdmission(
            handle: Long,
            secret: ByteBuffer,
            secretLength: Int,
            epoch: Long,
            maxAttempts: Int,
        ): Int
        @JvmStatic private external fun nativeIssueReplayToken(
            handle: Long,
            nowSeconds: Long,
            ttlSeconds: Long,
            output: ByteBuffer,
            limit: Int,
        ): Long
        @JvmStatic private external fun nativeRead(handle: Long, flow: Long, output: ByteBuffer, limit: Int): Long
        @JvmStatic private external fun nativeWrite(handle: Long, flow: Long, input: ByteBuffer, length: Int): Long
        @JvmStatic private external fun nativeFlush(handle: Long, flow: Long): Int
        @JvmStatic private external fun nativeShutdown(handle: Long, flow: Long): Int
        @JvmStatic private external fun nativeCloseFlow(handle: Long, flow: Long): Int
        @JvmStatic private external fun nativeClose(handle: Long): Int

        private fun status(packed: Long) = QuicpStatus.decode(packed.toInt())
        private fun status(raw: Int) = QuicpStatus.decode(raw)
        private fun value(packed: Long) = (packed ushr 32).toInt()
        private fun flow(rawStatus: Int, output: LongArray, engine: QuicpEngine): Result<QuicpFlow> {
            val decoded = status(rawStatus)
            return if (decoded == QuicpStatus.OK) {
                Result.success(QuicpFlow(engine, output[0]))
            } else {
                Result.failure(QuicpException(decoded))
            }
        }
    }

    val connectionStatus: QuicpStatus
        get() = if (handle == 0L) QuicpStatus.CLOSED else status(nativeConnectionState(handle))
    val nextTimerNanos: Long? get() = nativeNextTimer(handle).takeIf { it >= 0 }
    private val requestHost = ByteBuffer.allocateDirect(253)
    private val requestInitial = ByteBuffer.allocateDirect(32 * 1024)
    private val requestMetadata = LongArray(4)

    fun recoverySnapshot(): Result<QuicpRecoverySnapshot> {
        if (handle == 0L) return Result.failure(QuicpException(QuicpStatus.CLOSED))
        val values = LongArray(13)
        val status = status(nativeRecoverySnapshot(handle, values))
        return if (status == QuicpStatus.OK) {
            Result.success(
                QuicpRecoverySnapshot(
                    values[0], values[1], values[2], values[3], values[4],
                    values[5], values[6], values[7], values[8], values[9],
                    values[10], values[11], values[12],
                )
            )
        } else {
            Result.failure(QuicpException(status))
        }
    }

    fun drive(elapsedNanos: Long, maxTasks: Int = 256): QuicpDriveResult {
        if (handle == 0L) return QuicpDriveResult(QuicpStatus.CLOSED, 0)
        val packed = nativeDrive(handle, elapsedNanos, maxTasks)
        return QuicpDriveResult(status(packed), value(packed))
    }

    fun ingress(path: Int, data: ByteBuffer, length: Int): QuicpStatus {
        requireDirect(data)
        require(length in 1..data.limit())
        return if (handle == 0L) QuicpStatus.CLOSED else status(nativeIngress(handle, path, data, length))
    }

    fun egress(path: Int, output: ByteBuffer): QuicpIoResult {
        requireDirect(output, writable = true)
        if (handle == 0L) return QuicpIoResult(QuicpStatus.CLOSED, 0)
        val packed = nativeEgress(handle, path, output, output.limit())
        return QuicpIoResult(status(packed), value(packed))
    }

    fun markPathUnavailable(path: Int): QuicpStatus =
        if (handle == 0L) QuicpStatus.CLOSED else status(nativePathUnavailable(handle, path))

    fun openFlow(host: String, port: Int): Result<QuicpFlow> {
        require(port in 1..65535)
        if (handle == 0L) return Result.failure(QuicpException(QuicpStatus.CLOSED))
        val output = LongArray(1)
        return flow(nativeOpenFlow(handle, host.encodeToByteArray(), port, output), output, this)
    }

    fun openReplaySafeFlow(
        token: ByteBuffer,
        tokenLength: Int,
        nonce: Long,
        host: String,
        port: Int,
        initial: ByteBuffer,
        initialLength: Int,
    ): Result<QuicpFlow> {
        requireDirect(token)
        requireDirect(initial)
        require(tokenLength in 1..token.limit())
        require(initialLength in 1..initial.limit())
        require(port in 1..65535)
        if (handle == 0L) return Result.failure(QuicpException(QuicpStatus.CLOSED))
        val output = LongArray(1)
        val status = nativeOpenReplaySafeFlow(
            handle, token, tokenLength, nonce, host.encodeToByteArray(), port, initial,
            initialLength, output,
        )
        return flow(status, output, this)
    }

    fun pollFlowRequest(replaySafe: Boolean = false): Result<QuicpPendingFlow> {
        if (handle == 0L) return Result.failure(QuicpException(QuicpStatus.CLOSED))
        val status = status(
            nativePollFlowRequest(
                handle, replaySafe, requestHost, requestInitial, requestMetadata,
            )
        )
        if (status != QuicpStatus.OK) return Result.failure(QuicpException(status))
        val hostLength = requestMetadata[1].toInt()
        val initialLength = requestMetadata[3].toInt()
        if (hostLength !in 1..requestHost.capacity() || initialLength !in 0..requestInitial.capacity()) {
            return Result.failure(QuicpException(QuicpStatus.FAILED))
        }
        val host = ByteArray(hostLength)
        requestHost.position(0)
        requestHost.get(host)
        requestHost.clear()
        val initial = ByteArray(initialLength)
        requestInitial.position(0)
        requestInitial.get(initial)
        requestInitial.clear()
        return Result.success(
            QuicpPendingFlow(
                this,
                requestMetadata[0],
                host.decodeToString(throwOnInvalidSequence = true),
                requestMetadata[2].toInt(),
                initial,
            )
        )
    }

    fun configureReplayAdmission(
        secret: ByteBuffer,
        secretLength: Int,
        epoch: Long,
        maxAttempts: Int,
    ): QuicpStatus {
        requireDirect(secret)
        require(secretLength in 32..secret.limit())
        require(maxAttempts > 0)
        return if (handle == 0L) QuicpStatus.CLOSED else status(
            nativeConfigureReplayAdmission(handle, secret, secretLength, epoch, maxAttempts)
        )
    }

    fun issueReplayToken(
        nowSeconds: Long,
        ttlSeconds: Long,
        output: ByteBuffer,
    ): QuicpIoResult {
        require(nowSeconds >= 0 && ttlSeconds > 0)
        requireDirect(output, writable = true)
        if (handle == 0L) return QuicpIoResult(QuicpStatus.CLOSED, 0)
        val packed = nativeIssueReplayToken(handle, nowSeconds, ttlSeconds, output, output.limit())
        return QuicpIoResult(status(packed), value(packed))
    }

    internal fun read(flow: Long, output: ByteBuffer): QuicpIoResult {
        requireDirect(output, writable = true)
        if (handle == 0L) return QuicpIoResult(QuicpStatus.CLOSED, 0)
        val packed = nativeRead(handle, flow, output, output.limit())
        return QuicpIoResult(status(packed), value(packed))
    }

    internal fun write(flow: Long, input: ByteBuffer, length: Int): QuicpIoResult {
        requireDirect(input)
        require(length in 1..input.limit())
        if (handle == 0L) return QuicpIoResult(QuicpStatus.CLOSED, 0)
        val packed = nativeWrite(handle, flow, input, length)
        return QuicpIoResult(status(packed), value(packed))
    }

    internal fun flush(flow: Long) = if (handle == 0L) QuicpStatus.CLOSED else status(nativeFlush(handle, flow))
    internal fun shutdown(flow: Long) = if (handle == 0L) QuicpStatus.CLOSED else status(nativeShutdown(handle, flow))
    internal fun closeFlow(flow: Long) = if (handle == 0L) QuicpStatus.CLOSED else status(nativeCloseFlow(handle, flow))
    internal fun acceptPending(request: Long): Result<QuicpFlow> {
        if (handle == 0L) return Result.failure(QuicpException(QuicpStatus.CLOSED))
        val output = LongArray(1)
        return flow(nativeAcceptPendingFlow(handle, request, output), output, this)
    }
    internal fun rejectPending(request: Long) =
        if (handle == 0L) QuicpStatus.CLOSED else status(nativeRejectPendingFlow(handle, request))

    override fun close() {
        val current = handle
        if (current == 0L) return
        handle = 0
        nativeClose(current)
    }

    private fun requireDirect(buffer: ByteBuffer, writable: Boolean = false) {
        require(buffer.isDirect && buffer.position() == 0)
        if (writable) require(!buffer.isReadOnly)
    }
}

class QuicpPendingFlow internal constructor(
    private val engine: QuicpEngine,
    private var request: Long,
    val host: String,
    val port: Int,
    val initialData: ByteArray,
) {
    fun accept(): Result<QuicpFlow> {
        if (request == 0L) return Result.failure(QuicpException(QuicpStatus.CLOSED))
        val result = engine.acceptPending(request)
        if (result.isSuccess) request = 0
        return result
    }

    fun reject(): QuicpStatus {
        if (request == 0L) return QuicpStatus.CLOSED
        val status = engine.rejectPending(request)
        if (status == QuicpStatus.OK) request = 0
        return status
    }
}

class QuicpFlow internal constructor(private val engine: QuicpEngine, private var handle: Long) : AutoCloseable {
    fun read(output: ByteBuffer) = if (handle == 0L) QuicpIoResult(QuicpStatus.CLOSED, 0) else engine.read(handle, output)
    fun write(input: ByteBuffer, length: Int) = if (handle == 0L) QuicpIoResult(QuicpStatus.CLOSED, 0) else engine.write(handle, input, length)
    fun flush() = if (handle == 0L) QuicpStatus.CLOSED else engine.flush(handle)
    fun shutdown() = if (handle == 0L) QuicpStatus.CLOSED else engine.shutdown(handle)

    override fun close() {
        val current = handle
        if (current == 0L) return
        handle = 0
        engine.closeFlow(current)
    }
}

class QuicpException(val status: QuicpStatus) : Exception("QUICP status $status")

File: sdk/apple/Sources/Quicp/QuicpEngine.swift
import CQuicp
import Darwin

public enum QuicpStatus: UInt32, Error, Sendable {
  case ok = 0
  case wouldBlock = 1
  case bufferTooSmall = 2
  case invalidArgument = 3
  case notReady = 4
  case closed = 5
  case panic = 6
  case failed = 7

  fileprivate static func fromNative(_ value: UInt32) -> Self {
    Self(rawValue: value) ?? .panic
  }
}

public struct QuicpSocketAddress: Sendable {
  public let address: String
  public let port: UInt16

  public init(_ address: String, port: UInt16) {
    self.address = address
    self.port = port
  }

  fileprivate func native() throws -> quicp_socket_address_t {
    var result = quicp_socket_address_t()
    result.port = port
    if let ipv4 = IPv4Address(address) {
      result.family = 4
      withUnsafeMutableBytes(of: &result.address) { $0.copyBytes(from: ipv4.bytes) }
    } else if let ipv6 = IPv6Address(address) {
      result.family = 6
      withUnsafeMutableBytes(of: &result.address) { $0.copyBytes(from: ipv6.bytes) }
    } else {
      throw QuicpStatus.invalidArgument
    }
    return result
  }
}

private struct IPv4Address {
  let bytes: [UInt8]
  init?(_ value: String) {
    let octets = value.split(separator: ".", omittingEmptySubsequences: false)
    guard octets.count == 4 else { return nil }
    var bytes = [UInt8]()
    bytes.reserveCapacity(4)
    for octet in octets {
      guard !octet.isEmpty,
            octet == "0" || octet.first != "0",
            octet.allSatisfy({ $0.isASCII && $0.isNumber }),
            let byte = UInt8(octet)
      else { return nil }
      bytes.append(byte)
    }
    self.bytes = bytes
  }
}

private struct IPv6Address {
  let bytes: [UInt8]
  init?(_ value: String) {
    var storage = in6_addr()
    guard value.withCString({ inet_pton(AF_INET6, $0, &storage) }) == 1 else { return nil }
    self.bytes = withUnsafeBytes(of: storage) { Array($0) }
  }
}

public struct QuicpPath: Sendable {
  public let local: QuicpSocketAddress
  public let peer: QuicpSocketAddress

  public init(local: QuicpSocketAddress, peer: QuicpSocketAddress) {
    self.local = local
    self.peer = peer
  }
}

public struct QuicpRecoverySnapshot: Sendable, Equatable {
  public let sourceSent: UInt64
  public let sourceReceived: UInt64
  public let repairSent: UInt64
  public let recovered: UInt64
  public let replayed: UInt64
  public let fallback: UInt64
  public let dropped: UInt64
  public let earlyAccepted: UInt64
  public let earlyRejected: UInt64
  public let pathLostPackets: UInt64
  public let maxPathRttMicros: UInt64
  public let queuedDatagrams: UInt64
  public let retainedSourceBytes: UInt64

  fileprivate init(_ value: quicp_recovery_snapshot_t) {
    sourceSent = value.source_sent
    sourceReceived = value.source_received
    repairSent = value.repair_sent
    recovered = value.recovered
    replayed = value.replayed
    fallback = value.fallback
    dropped = value.dropped
    earlyAccepted = value.early_accepted
    earlyRejected = value.early_rejected
    pathLostPackets = value.path_lost_packets
    maxPathRttMicros = value.max_path_rtt_micros
    queuedDatagrams = value.queued_datagrams
    retainedSourceBytes = value.retained_source_bytes
  }
}

public final class QuicpEngine {
  public enum Role: UInt32, Sendable { case client = 1, server = 2 }
  public enum RecoveryMode: UInt32, Sendable { case adaptive = 1, reliableOnly = 2 }
  public enum Security: Sendable {
    case insecure
    case mutualTLS(
      serverName: String,
      caCertificate: String,
      certificate: String,
      privateKey: String
    )
  }

  private var raw: OpaquePointer?

  public init(
    role: Role,
    paths: [QuicpPath],
    packetCapacity: UInt32 = 256,
    mtu: UInt32 = 1500,
    recoveryMode: RecoveryMode = .adaptive,
    security: Security = .insecure
  ) throws {
    guard quicp_abi_version() == QUICP_ABI_VERSION, (1...2).contains(paths.count) else {
      throw QuicpStatus.invalidArgument
    }
    let first = quicp_path_config_t(
      local: try paths[0].local.native(),
      peer: try paths[0].peer.native()
    )
    let second = paths.count == 2
      ? quicp_path_config_t(local: try paths[1].local.native(), peer: try paths[1].peer.native())
      : quicp_path_config_t()
    var config = quicp_engine_config_t(
      abi_version: QUICP_ABI_VERSION,
      role: role.rawValue,
      path_count: UInt32(paths.count),
      paths: (first, second),
      packet_capacity: packetCapacity,
      mtu: mtu,
      recovery_mode: recoveryMode.rawValue
    )
    let nativeStatus: UInt32
    switch security {
    case .insecure:
      nativeStatus = quicp_engine_create(&config, &raw)
    case let .mutualTLS(serverName, caCertificate, certificate, privateKey):
      let values = [serverName, caCertificate, certificate, privateKey].map { Array($0.utf8) }
      nativeStatus = values[0].withUnsafeBufferPointer { serverName in
        values[1].withUnsafeBufferPointer { caCertificate in
          values[2].withUnsafeBufferPointer { certificate in
            values[3].withUnsafeBufferPointer { privateKey in
              var tls = quicp_tls_config_t(
                server_name: quicp_bytes_t(
                  data: serverName.baseAddress, length: UInt32(serverName.count)
                ),
                ca_certificate: quicp_bytes_t(
                  data: caCertificate.baseAddress, length: UInt32(caCertificate.count)
                ),
                certificate: quicp_bytes_t(
                  data: certificate.baseAddress, length: UInt32(certificate.count)
                ),
                private_key: quicp_bytes_t(
                  data: privateKey.baseAddress, length: UInt32(privateKey.count)
                )
              )
              return quicp_engine_create_tls(&config, &tls, &raw)
            }
          }
        }
      }
    }
    let status = QuicpStatus.fromNative(nativeStatus)
    guard status == .ok else { throw status }
  }

  deinit { _ = close() }

  public var connectionStatus: QuicpStatus {
    guard let raw else { return .closed }
    return QuicpStatus.fromNative(quicp_engine_connection_state(raw))
  }

  @discardableResult
  public func drive(elapsedNanoseconds: UInt64, maxTasks: UInt32 = 256) -> (QuicpStatus, UInt32) {
    guard let raw else { return (.closed, 0) }
    var processed: UInt32 = 0
    let status = quicp_engine_drive(raw, elapsedNanoseconds, maxTasks, &processed)
    return (QuicpStatus.fromNative(status), processed)
  }

  public var nextTimerNanoseconds: UInt64? {
    guard let raw else { return nil }
    var present: UInt32 = 0
    var deadline: UInt64 = 0
    guard quicp_engine_next_timer(raw, &present, &deadline) == QUICP_STATUS_OK else { return nil }
    return present == 0 ? nil : deadline
  }

  public var recoverySnapshot: Result<QuicpRecoverySnapshot, QuicpStatus> {
    guard let raw else { return .failure(.closed) }
    var snapshot = quicp_recovery_snapshot_t()
    let status = QuicpStatus.fromNative(quicp_engine_recovery_snapshot(raw, &snapshot))
    return status == .ok ? .success(QuicpRecoverySnapshot(snapshot)) : .failure(status)
  }

  public func ingress(path: UInt32, datagram: UnsafeRawBufferPointer) -> QuicpStatus {
    guard let raw, let base = datagram.baseAddress else { return .invalidArgument }
    return QuicpStatus.fromNative(quicp_engine_ingress(
      raw, path, base.assumingMemoryBound(to: UInt8.self), UInt32(datagram.count)
    ))
  }

  public func egress(path: UInt32, into output: UnsafeMutableRawBufferPointer) -> (QuicpStatus, Int) {
    guard let raw, let base = output.baseAddress else { return (.invalidArgument, 0) }
    var length: UInt32 = 0
    let status = quicp_engine_egress(
      raw, path, base.assumingMemoryBound(to: UInt8.self), UInt32(output.count), &length
    )
    return (QuicpStatus.fromNative(status), Int(length))
  }

  public func markPathUnavailable(_ path: UInt32) -> QuicpStatus {
    guard let raw else { return .closed }
    return QuicpStatus.fromNative(quicp_engine_path_unavailable(raw, path))
  }

  public func openFlow(host: String, port: UInt16) -> Result<QuicpFlow, QuicpStatus> {
    guard let raw else { return .failure(.closed) }
    var handle: UInt64 = 0
    let status = host.utf8CString.withUnsafeBufferPointer { bytes in
      quicp_engine_open_flow(raw, bytes.baseAddress, UInt32(bytes.count - 1), port, &handle)
    }
    let value = QuicpStatus.fromNative(status)
    return value == .ok ? .success(QuicpFlow(engine: self, handle: handle)) : .failure(value)
  }

  public func openReplaySafeFlow(
    token: UnsafeRawBufferPointer,
    nonce: UInt64,
    host: String,
    port: UInt16,
    initial: UnsafeRawBufferPointer
  ) -> Result<QuicpFlow, QuicpStatus> {
    guard let raw,
          let tokenBase = token.baseAddress,
          let initialBase = initial.baseAddress
    else { return .failure(.invalidArgument) }
    var handle: UInt64 = 0
    let status = host.utf8CString.withUnsafeBufferPointer { bytes in
      quicp_engine_open_replay_safe_flow(
        raw,
        tokenBase.assumingMemoryBound(to: UInt8.self), UInt32(token.count), nonce,
        bytes.baseAddress, UInt32(bytes.count - 1), port,
        initialBase.assumingMemoryBound(to: UInt8.self), UInt32(initial.count),
        &handle
      )
    }
    let value = QuicpStatus.fromNative(status)
    return value == .ok ? .success(QuicpFlow(engine: self, handle: handle)) : .failure(value)
  }

  public func pollFlowRequest(replaySafe: Bool = false) -> Result<QuicpPendingFlow, QuicpStatus> {
    guard let raw else { return .failure(.closed) }
    var request: UInt64 = 0
    var host = [UInt8](repeating: 0, count: Int(QUICP_MAX_HOST_BYTES))
    var hostLength: UInt32 = 0
    var port: UInt16 = 0
    var initial = [UInt8](repeating: 0, count: Int(QUICP_MAX_EARLY_INITIAL_BYTES))
    var initialLength: UInt32 = 0
    let rawStatus = host.withUnsafeMutableBufferPointer { host in
      initial.withUnsafeMutableBufferPointer { initial in
        if replaySafe {
          return quicp_engine_poll_replay_safe_flow_request(
            raw, &request,
            host.baseAddress, UInt32(host.count), &hostLength, &port,
            initial.baseAddress, UInt32(initial.count), &initialLength
          )
        }
        return quicp_engine_poll_flow_request(
          raw, &request,
          host.baseAddress, UInt32(host.count), &hostLength, &port,
          initial.baseAddress, UInt32(initial.count), &initialLength
        )
      }
    }
    let status = QuicpStatus.fromNative(rawStatus)
    guard status == .ok else { return .failure(status) }
    let name = String(decoding: host.prefix(Int(hostLength)), as: UTF8.self)
    return .success(QuicpPendingFlow(
      engine: self,
      request: request,
      host: name,
      port: port,
      initialData: Array(initial.prefix(Int(initialLength)))
    ))
  }

  public func configureReplayAdmission(
    secret: UnsafeRawBufferPointer,
    epoch: UInt64,
    maxAttempts: UInt32
  ) -> QuicpStatus {
    guard let raw, let base = secret.baseAddress else { return .invalidArgument }
    return QuicpStatus.fromNative(quicp_engine_configure_replay_admission(
      raw, base.assumingMemoryBound(to: UInt8.self), UInt32(secret.count), epoch, maxAttempts
    ))
  }

  public func issueReplayToken(
    nowSeconds: UInt64,
    ttlSeconds: UInt64,
    into output: UnsafeMutableRawBufferPointer
  ) -> (QuicpStatus, Int) {
    guard let raw else { return (.closed, 0) }
    var length: UInt32 = 0
    let status = quicp_engine_issue_replay_token(
      raw, nowSeconds, ttlSeconds,
      output.baseAddress?.assumingMemoryBound(to: UInt8.self),
      UInt32(output.count), &length
    )
    return (QuicpStatus.fromNative(status), Int(length))
  }

  @discardableResult
  public func close() -> QuicpStatus {
    guard raw != nil else { return .closed }
    return QuicpStatus.fromNative(quicp_engine_close(&raw))
  }

  fileprivate func read(_ handle: UInt64, into output: UnsafeMutableRawBufferPointer) -> (QuicpStatus, Int) {
    guard let raw, let base = output.baseAddress else { return (.invalidArgument, 0) }
    var count: UInt32 = 0
    let status = quicp_flow_read(
      raw, handle, base.assumingMemoryBound(to: UInt8.self), UInt32(output.count), &count
    )
    return (QuicpStatus.fromNative(status), Int(count))
  }

  fileprivate func write(_ handle: UInt64, from input: UnsafeRawBufferPointer) -> (QuicpStatus, Int) {
    guard let raw, let base = input.baseAddress else { return (.invalidArgument, 0) }
    var count: UInt32 = 0
    let status = quicp_flow_write(
      raw, handle, base.assumingMemoryBound(to: UInt8.self), UInt32(input.count), &count
    )
    return (QuicpStatus.fromNative(status), Int(count))
  }

  fileprivate func flush(_ handle: UInt64) -> QuicpStatus {
    guard let raw else { return .closed }
    return QuicpStatus.fromNative(quicp_flow_flush(raw, handle))
  }

  fileprivate func shutdown(_ handle: UInt64) -> QuicpStatus {
    guard let raw else { return .closed }
    return QuicpStatus.fromNative(quicp_flow_shutdown(raw, handle))
  }

  fileprivate func closeFlow(_ handle: UInt64) -> QuicpStatus {
    guard let raw else { return .closed }
    return QuicpStatus.fromNative(quicp_flow_close(raw, handle))
  }

  fileprivate func accept(_ request: UInt64) -> Result<QuicpFlow, QuicpStatus> {
    guard let raw else { return .failure(.closed) }
    var handle: UInt64 = 0
    let status = QuicpStatus.fromNative(
      quicp_engine_accept_pending_flow(raw, request, &handle)
    )
    return status == .ok ? .success(QuicpFlow(engine: self, handle: handle)) : .failure(status)
  }

  fileprivate func reject(_ request: UInt64) -> QuicpStatus {
    guard let raw else { return .closed }
    return QuicpStatus.fromNative(quicp_engine_reject_pending_flow(raw, request))
  }
}

public final class QuicpPendingFlow {
  private let engine: QuicpEngine
  private var request: UInt64
  public let host: String
  public let port: UInt16
  public let initialData: [UInt8]

  fileprivate init(
    engine: QuicpEngine,
    request: UInt64,
    host: String,
    port: UInt16,
    initialData: [UInt8]
  ) {
    self.engine = engine
    self.request = request
    self.host = host
    self.port = port
    self.initialData = initialData
  }

  public func accept() -> Result<QuicpFlow, QuicpStatus> {
    guard request != 0 else { return .failure(.closed) }
    let result = engine.accept(request)
    if case .success = result { request = 0 }
    return result
  }

  public func reject() -> QuicpStatus {
    guard request != 0 else { return .closed }
    let status = engine.reject(request)
    if status == .ok { request = 0 }
    return status
  }
}

public final class QuicpFlow {
  private let engine: QuicpEngine
  private var handle: UInt64

  fileprivate init(engine: QuicpEngine, handle: UInt64) {
    self.engine = engine
    self.handle = handle
  }

  deinit { _ = close() }

  public func read(into output: UnsafeMutableRawBufferPointer) -> (QuicpStatus, Int) {
    handle == 0 ? (.closed, 0) : engine.read(handle, into: output)
  }

  public func write(from input: UnsafeRawBufferPointer) -> (QuicpStatus, Int) {
    handle == 0 ? (.closed, 0) : engine.write(handle, from: input)
  }

  public func flush() -> QuicpStatus {
    handle == 0 ? .closed : engine.flush(handle)
  }

  public func shutdown() -> QuicpStatus {
    handle == 0 ? .closed : engine.shutdown(handle)
  }

  @discardableResult
  public func close() -> QuicpStatus {
    guard handle != 0 else { return .closed }
    let current = handle
    handle = 0
    return engine.closeFlow(current)
  }
}
